apps/labs/posts/building-scipy-with-flang.md

```diff
@@ -0,0 +1,598 @@
+
+## Python packaging
+
+This situation also happens to be _the_ fundamental problem in Python packaging.
+There's so much complexity hiding under the surface, that actually dealing with
+it in any kind of sane way is really difficult.
+For example, the C & C++ ecosystems never managed to standardise any sort of
+build & distribution mechanism, and since Python is wrapping around a lot of
+C & C++ code, the problems in Python packaging are a _superset_ of what's
+already a hugely pressing issue in C/C++.
+
+The packaging world in Python has had to continually reinvent itself in order
+to solve these issues.
+Here's a good [historical](https://www.youtube.com/watch?v=AQsZsgJ30AE&t=200s)
+overview of the many iterations this had gone through.
+One of the biggest problems that did get solved through the "wheel" format,
+was to avoid users having to compile anything to install a package.
+
+!['A meme about a difficult choice in Python packaging, with a man unable to decide between two buttons labeled "ABI Hell" and "Users need a compiler"'](/posts/building-scipy-with-flang/ABI_meme.jpg)
+
+This again comes with substantial maintenance trade-offs, because there are
+still many very involved things that maintainers have to take care of when
+preparing a wheel for their users.
+In terms of the necessary tooling, there was a point in this long evolution
+where `distutils` emerged as the "one way" to solve things, so much so that it
+even got adopted into the standard library.
+
+However, this was not enough for the Scientific Python community, which had
+bigger problems.
+This [blog](http://technicaldiscovery.blogspot.com/2013/12/why-i-promote-conda.html)
+by one of the founding contributors of NumPy and SciPy is full of relevant
+anecdotes; here's one quote:
+
+> We in the scientific python community have had difficulty and a rocky history
+> with just waiting for the Python.org community to solve the [packaging]
+> problem. With `distutils` for example, we had to essentially re-write most of
+> it (as `numpy.distutils`) in order to support compilation of extensions that
+> needed Fortran-compiled libraries.
+
+Despite being the product of heroic efforts by many very bright people,
+`numpy.distutils` is generally regarded as a pile of hacks, and has been in
+minimal maintenance mode for years.
+In many ways, this is what led to the creation of conda.
+Again from that blog post:
+
+> Therefore, you can't really address the problem of Python packaging without
+> addressing the core problems of trying to use `distutils` (at least for the
+> NumPy stack). The problems for us in the NumPy stack started there and have
+> to be rooted out there as well. This was confirmed for me at the first PyData
+> meetup at Google HQ, where several of us asked Guido [van Rossum; Python's
+>
+> > BDFL at the time] what we can do to fix Python packaging for the NumPy stack.
+> > Guido's answer was to "solve the problem ourselves".
```

Comment:
Terrific question, @h-vetinari! That looks like an autoformatter goof. I should've reviewed those more closely. 😬 

```suggestion
> meetup at Google HQ, where several of us asked Guido [van Rossum; Python's
> BDFL at the time] what we can do to fix Python packaging for the NumPy stack.
> Guido's answer was to "solve the problem ourselves".
```